### PR TITLE
[ViewPager] Add setPageWithoutAnimation

### PR DESCRIPTION
**Pull request by brentvatne**
In some cases it's desirable to set the page in the ViewPager without animating it -- we have this in ScrollView with `scrollWithoutAnimationTo`. This PR adds `setPageWithoutAnimation` on ViewPager.

cc @ide @kmagiera

### CODE REVIEW

**Comment by mkonicek** [comment]
Thanks Brent! Will take a look.

File: ReactAndroid/src/main/java/com/facebook/react/views/viewpager/ReactViewPagerManager.java [hunk 1]
```diff
@@ -41,6 +41,11 @@ public void setSelectedPage(ReactViewPager view, int page) {
     view.setCurrentItemFromJs(page);
   }
 
+  @ReactProp(name = "animatePageTransition")
+  public void setAnimatePageTransition(ReactViewPager view, boolean animatedPageTransition) {
```
Comment (mkonicek) [inline ReactAndroid/src/main/java/com/facebook/react/views/viewpager/ReactViewPagerManager.java]: nit: `animated` -> `animate`

File: ReactAndroid/src/main/java/com/facebook/react/views/viewpager/ReactViewPagerManager.java [hunk 2]
```diff
@@ -58,4 +63,10 @@ public Map getExportedCustomDirectEventTypeConstants() {
   public void addView(ReactViewPager parent, View child, int index) {
     parent.addViewToAdapter(child, index);
   }
+
+  @Override
+  protected void onAfterUpdateTransaction(ReactViewPager view) {
```
Comment (mkonicek) [inline ReactAndroid/src/main/java/com/facebook/react/views/viewpager/ReactViewPagerManager.java]: Why is this needed? I'm not very familiar with the ViewPager API. Can you add a comment?

File: Libraries/Components/ViewPager/ViewPagerAndroid.android.js [hunk 3]
```diff
@@ -151,6 +152,13 @@ var ViewPagerAndroid = React.createClass({
   setPage: function(selectedPage) {
     this.setState({
       selectedPage,
+      animatePageTransition: true,
```
Comment (mkonicek) [inline Libraries/Components/ViewPager/ViewPagerAndroid.android.js]: This has a hidden side effect of setting `animatePageTransition` for all future updates of `selectedPage`.

How about making this explicit and exposing a `animateSetPage` prop? (Feel free to come up with a better name :))

File: ReactAndroid/src/main/java/com/facebook/react/views/viewpager/ReactViewPagerManager.java [hunk 4]
```diff
@@ -41,6 +41,11 @@ public void setSelectedPage(ReactViewPager view, int page) {
     view.setCurrentItemFromJs(page);
```
Comment (mkonicek) [inline ReactAndroid/src/main/java/com/facebook/react/views/viewpager/ReactViewPagerManager.java]: Is the comment above this line still valid? The task only has a link to this file :) It was added by Krzysztof.

File: Libraries/Components/ViewPager/ViewPagerAndroid.android.js (hunk 3 above)
Comment (brentvatne) [inline Libraries/Components/ViewPager/ViewPagerAndroid.android.js]: > This has a hidden side effect of setting animatePageTransition for all future updates of selectedPage.

It actually doesn't have any impact because this only happens for JS-driven page updates, and those always go through `setPage` or `setPageWithoutAnimation` so we set the animated transition prop each time. It's a little bit of an awkward API though I agree. It leads to the `onAfterUpdateTransaction` method above as well because we need to be sure that all of the prop updates for the batch have been applied before actually setting the page on the ViewPager.

File: ReactAndroid/src/main/java/com/facebook/react/views/viewpager/ReactViewPagerManager.java (hunk 2 above)
Comment (brentvatne) [inline ReactAndroid/src/main/java/com/facebook/react/views/viewpager/ReactViewPagerManager.java]: See my reply below! :)

File: ReactAndroid/src/main/java/com/facebook/react/views/viewpager/ReactViewPagerManager.java (hunk 1 above)
Comment (brentvatne) [inline ReactAndroid/src/main/java/com/facebook/react/views/viewpager/ReactViewPagerManager.java]: Going to get rid of this as per comment below :)

**Comment by brentvatne** [comment]
@mkonicek - updated!

**Comment by mkonicek** [comment]
Oh sorry I meant:

    // All page updates are instant
    <ViewPager animateSetPage={false} selectedPage={this.state.page} ... />

    // All page updates are animated
    <ViewPager selectedPage={this.state.page} ... />

Do you think that API would make sense?

It's similar to how you set the `bounces` property on a ScrollView - it's a prop to control its visual behavior.

With the current solution there are two ways to set a page which seems a bit confusing to me: logically a ViewPager only has one selected page, why does it also have a `selectedPageWithoutAnimation`?

**Comment by mkonicek** [comment]
Is it sufficient for you to create a ViewPager that never animates, or do you need to flip between animated and instant page sets dynamically?

The dynamic case is harder: I think we should have some way to set both `animateSetPage` and `selectedPage` in the same batch (same `render()`) and have the native side set all the props first and only then call `setCurrentItemFromJs`.

**Comment by mkonicek** [comment]
Just talked to Krzysztof who made a really good point - it might be better to have method `setPage`, `setPageWithoutAnimation`, same as `ScrollView.scrollTo`, `ScrollView.scrollToWithoutAnimation`. But then, it's not the "React" way :)

We think this requires more discussion, will chat about it with more people at fb to get more opinions.

**Comment by brentvatne** [comment]
> Just talked to Krzysztof who made a really good point - another option is not to have a selectedPage prop at all but have methods `setPage`, `setPageWithoutAnimation`. But then, is it the right API from React's perspective?

This is the exact approach that we are currently using with this PR -- the public interface for props only exposes `initialPage`, `onPageScroll`, `onPageSelected`, and `keyboardDismissMode`. In addition to this, we expose `setPage` or `setPageWithoutAnimation` functions that can be called on the ref.

Internally, when `setPage` or `setPageWithoutAnimation` are called, we use props to update those. I think this part is an implementation detail and I just went along with how `setPage` was implemented already, but if we wanted some more consistency with how it's done on iOS at the moment then we could look at ScrollView and the equivalents `scrollTo` and `scrollWithoutAnimationTo`. These rely on bridge methods on `RCTUIManager` of the same name. So the implementation of `setPage(selectedPage)` on `ViewPagerAndroid.android.js` would look something like this:

```
  setPage: function(selectedPage) {
    ViewPagerManager.setPage(selectedPage);
  },
```

rather than its current implementation:

```
  setPage: function(selectedPage) {
    this.setState({
      selectedPage,
    });
  },
```

**Comment by brentvatne** [comment]
@mkonicek - ok thinking about this some more and I think the currently implemented approach of using props to update the page is flawed. Think of the following situation:

- You update the page using `setPage`
- You then scroll a bit manually
- Then you do something to update the page using `setPage` but you go back to the previous page that you were at before.

This will *not* work because when React does the prop diff it will see that the props are the same. So, I think for values that live on the native side like this and scroll position and such, we either need to have a very good approach for keeping that in sync with the JS side or just use the approach from ScrollView to set them imperatively.

**Comment by machard** [comment]
I need this feature too to make a looping carousel.
I looked quickly how to code the imperative approach but I'm not sure how to access the view object from a ReactMethod. I think I should use the reactTag but I'm not sure what to use to get the view from it.

**Comment by mkonicek** [comment]
@brentvatne I realized we also want to make the `ViewPagerAndroid` API closer to the iOS `ScrollView` for consistency: We might even want to render a `ViewPagerAndroid` under the hood if you specify `<ScrollView pagingEnabled />`.

Therefore I think we should get rid of the prop and have a function to do the paging imperatively, like `ScrollView` does.

**Comment by mkonicek** [comment]
@machard Appreciate your help! You can read the code for `ScrollView.scrollTo` for example, this will be very similar. But I think @brentvatne has got this once he's back from vacation :)

**Comment by mkonicek** [comment]
@brentvatne Your example with setting the page to 5, scrolling and then setting it again is great! We should have thought of that when implementing this. Luckily it's not so hard to change it :)

**Comment by brentvatne** [comment]
Hiya @mkonicek!

> @brentvatne I realized we also want to make the ViewPagerAndroid API closer to the iOS ScrollView for consistency: We might even want to render a ViewPagerAndroid under the hood if you specify <ScrollView pagingEnabled />.

Yup! That would be nice to have eventually.

> Therefore I think we should get rid of the prop and have a function to do the paging imperatively, like ScrollView does.

> We should make this work exactly the same way as on iOS. It doesn't look like ScrollView has an easy way to scroll to a given page though?

Correct! Right now you have to just know the width of each page and multiply that by the page that you want to be on. Would be nice to have this on both though, agreed.

> We can make this consistent in steps:
> 1. First, change the ViewPagerAndroid API to have two imperative functions for scrolling, get rid of the  prop. Update the example. That's a good start, let's do that in this PR.
> 2. Add support for scrolling to page to iOS ScrollView
> 3. Unify both APIs

Sounds good! I took care of #1 here. `setPage` and `setPageWithoutAnimation` are implemented imperatively now.

**Comment by ide** [comment]
A brief thought on paginated scrollviews and ViewPager -- we may want (hypothetically speaking) ViewPager.ios.js to use a paginated ScrollView instead of having ScrollView sometimes render a ViewPager. I think of ScrollView as a lower-level component than ViewPager, and it might be easier to implement ViewPager on iOS using a ScrollView, than it is to implement a paginated ScrollView on Android using a ViewPager (think of all the scroll responder APIs).

**Comment by mkonicek** [comment]
That's a good point @ide.

@brentvatne Looks awesome! I think you might be able to inline the `ReactViewPagerCommandHelper` into the View pager, the reason we have that helper for ScrollView is that there is a vertical and horizontal one and wanted to share code between them.

Does the `initialPage` prop as expected with this change? Could you also try running and update the example please?
https://github.com/facebook/react-native/blob/master/Examples/UIExplorer/ViewPagerAndroidExample.android.js

**Comment by brentvatne** [comment]
> I think you might be able to inline the ReactViewPagerCommandHelper into the View pager, the reason we have that helper for ScrollView is that there is a vertical and horizontal one and wanted to share code between them.

Done!

> Does the initialPage prop as expected with this change? Could you also try running and update the example please?

It does now! Updated example to add a button to toggle between using `setPage` and `setPageWithoutAnimation`

File: Libraries/Components/ViewPager/ViewPagerAndroid.android.js [hunk 5]
```diff
@@ -97,17 +97,17 @@ var ViewPagerAndroid = React.createClass({
...
+    }
   },
 
-  getInnerViewNode: function() {
+  getInnerViewNode: function(): ReactComponent {
     return this.refs[VIEWPAGER_REF].getInnerViewNode();
   },
 
-  _childrenWithOverridenStyle: function() {
+  _childrenWithOverridenStyle: function(): Array {
```
Comment (brentvatne) [inline Libraries/Components/ViewPager/ViewPagerAndroid.android.js]: Not great with flow -- what should I put here as the return type @mkonicek?

Comment (mkonicek) [inline Libraries/Components/ViewPager/ViewPagerAndroid.android.js]: Good question. Try `Array<ReactElement>` and run `flow` from the root of the repo and see if that works.

Doesn't look like the React codebase uses Flow so not sure how the Flow handles the boundaries between code that uses Flow and code that doesn't use it.

If `Array<ReactElement>` doesn't work I'd say `Array` or `Array<Object>` are perfectly fine - it's a private function you're using just in this file.

**Comment by mkonicek** [comment]
Looks awesome! :+1: Thanks for improving the example and using Flow too!

Sorry for the delay. Feel free to try out what return type works best for `_childrenWithOverridenStyle` (or just use `Array`) and ping me when you want this merged :)

File: Libraries/Components/ViewPager/ViewPagerAndroid.android.js (hunk 5 above)
Comment (kmagiera) [inline Libraries/Components/ViewPager/ViewPagerAndroid.android.js]: It is no longer required to override styles this way here. We did this initially because we didn't have view flattening on android back then. Now that view flattening is there we should just wrap children in views positioned absolutely and with `collapsable=false` similarly to how we do it here: https://github.com/facebook/react-native/blob/master/Libraries/Components/ScrollView/RecyclerViewBackedScrollView.android.js#L95

Comment (mkonicek) [inline Libraries/Components/ViewPager/ViewPagerAndroid.android.js]: @kmagiera It's a good point and looked into doing that when open sourcing the view pager but decided not to (don't remember the exact reason anymore). I think changing this is outside the scope of Brent's PR though. I'll pull this in as it looks good otherwise.

**Comment by mkonicek** [comment]
@facebook-github-bot shipit
